fetcher.go

```diff
@@ -25,20 +23,19 @@ type FetcherConfig struct {
 
 type Fetcher interface {
 	// NodeMatching traverses a node graph starting with the provided node using the given selector and possibly crossing
-	// block boundaries. Each matched node is sent to the FetchResult channel.
-	// The results and error channels will be closed on query completion or error. The error channel is buffered,
-	// will emit at most one error and must be checked after processing the results.
-	NodeMatching(ctx context.Context, node ipld.Node, match selector.Selector) (<-chan FetchResult, <-chan error)
+	// block boundaries. Each matched node is sent to the FetchResult channel.  This method blocks until all results
+	// are read or an error occurs. The provided function will be called once for every result and possibly once with an
+	// error after which not be called.
+	NodeMatching(context.Context, ipld.Node, selector.Selector, FetchCallback)
```

Comment:
comment on semantics of when `FetchCallback` will be called vs when the method returns (e.g. does the method call block until all callbacks are called or is it expected that the call it self will return quickly, although the callback may happen later?)

Reply:
I wrote this out as a sequence of events to clarify that it blocks until all the callbacks have been called.

Reply:
if it blocks until all results, is it worth returning an error, and only having the callback for results?